src/charm.py

```diff
@@ -0,0 +1,205 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm the application."""
+
+import logging
+
+import ops
+from charms.data_platform_libs.v0.data_models import TypedCharmBase
+from charms.rolling_ops.v0.rollingops import RollingOpsManager
+from ops.pebble import Layer, LayerDict
+
+from core.cluster import ClusterContext
+from core.structured_config import CharmConfig
+from events.kafka import KafkaHandler
+from events.password_actions import PasswordActionEvents
+from events.provider import KarapaceHandler
+from events.tls import TLSHandler
+from literals import CHARM_KEY, CONTAINER, GROUP, USER, DebugLevel, Status, Substrate
+from managers.auth import KarapaceAuth
+from managers.config import ConfigManager
+from managers.kafka import KafkaManager
+from managers.tls import TLSManager
+from workload import KarapaceWorkload
+
+logger = logging.getLogger(__name__)
+
+
+class KarapaceCharm(TypedCharmBase[CharmConfig]):
+    """Charmed Operator for Karapace."""
+
+    config_type = CharmConfig
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.name = CHARM_KEY
+        self.substrate: Substrate = "k8s"
+        self.context = ClusterContext(charm=self, substrate=self.substrate)
+        self.workload = KarapaceWorkload(container=self.unit.get_container(CONTAINER))
+
+        # HANDLERS
+
+        self.password_action_events = PasswordActionEvents(self)
+        self.kafka = KafkaHandler(self)
+        self.tls = TLSHandler(self)
+        self.provider = KarapaceHandler(self)
+
+        # MANAGERS
+
+        self.config_manager = ConfigManager(context=self.context, workload=self.workload)
+        self.auth_manager = KarapaceAuth(context=self.context, workload=self.workload)
+        self.tls_manager = TLSManager(context=self.context, workload=self.workload)
+        self.kafka_manager = KafkaManager(context=self.context, workload=self.workload)
+
+        # LIB HANDLERS
+
+        self.restart = RollingOpsManager(self, relation="restart", callback=self._restart)
+
+        # CORE EVENTS
+
+        self.framework.observe(self.on.install, self._on_install)
+        # self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(
+            getattr(self.on, "karapace_pebble_ready"), self._on_karapace_pebble_ready
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+
+    @property
+    def karapace_layer(self) -> Layer:
```

Comment:
**todo** this seems out of place here, I believe it would be more appropriate in the workload implementation in `src/workload.py`. This would also allow to keep more consistency between VM and K8s